### ISSUE TITLE
[Q] Input lag approximation based on Present data

### ISSUE THREAD

**Issue by DevTechProfile** [GameTechDev/PresentMon#106]
First of all a [graphic from Nvidia](https://www.nvidia.com/en-us/geforce/guides/system-latency-optimization-guide/) to have a common terminology. 

![grafik](https://user-images.githubusercontent.com/30285858/124353458-ebdd5e80-dc06-11eb-9fe9-2913ce0195cc.png)

 What we offer in CapFrameX is an approximation of the _System Latency_ (input lag). _Periphal Latency_ and _Display Latency_ are represented via an (configurable) offset. The more exiciting part is the _PC Latency._

We are using (Upper bound + Lower bound) / 2 = expected latency, where

Lower bound = frameTimes[i] + untilDisplayedTimes[i] - inPresentAPITimes[i - 1] and
Upper bound = frameTimes[i] + untilDisplayedTimes[i] + frameTimes[i - 1] - inPresentAPITimes[i - 2].

Q1: Is this approach (expected latency) valid to provide a suitable _PC Latency_ approximation?

The next point is something about odd untilDisplayedTimes and inPresentAPITimes numbers. I've measured Doom Eternal and noticed in the capture data that inPresentAPITimes is always 0. Additionallly I noticed that untilDisplayedTimes is very high. The numbers were about 14-16ms. No frame limiter was used, VSync was disabled. 

On the other hand Stranded Deep (DX11) gave me inPresentAPITimes numbers within a range of 0.2-0.5ms and untilDisplayedTimes within a range of 2-3ms. As before, no frame limiter was used, VSync was disabled. 

Q2: Are these numbers correct? Why are there such big differences between Vulkan and DX11 titles? We actually tested more games which showed a very similar behavior. 

Q3: What is a correct way to handle dropped presents when calculating the _PC Latency_ approximation ? What we are currently doing is to filter out presents with a dropped flag = true.

**Comment by aggieNick02** [comment]
On Q1, Doom Eternal uses Vulkan, so inPresentAPITimes will always be 0, as noted in the issue at the bottom of the main page: https://github.com/GameTechDev/PresentMon . It seems odd that the msUntilDisplayed are high - as the note on the main page says, they should actually be artificially low, since the msUntilDisplayed is measured from TimeInSeconds, which is delayed by the amount of time really spent in present, since that isn't measurable for Vulkan/OpenGL.

The 14-16 ms for untilDisplayed does sound high; do you have a sample CSV?

On Q2, the note for OpenGL/Vulkan apps explains many of the differences you'll see between dx and non-dx titles

On Q3, I don't think filtering the dropped presents is correct. All of the time spent on those dropped frames is added latency between the presents not dropped before and after any batch of drops. If you have a present at row i with n dropped presents before it, then the formula on the main page becomes
`LatencyMs[i] =~ Sum(j=i-n,j=i,msBetweenPresents[j]) + msUntilDisplayed[i] - msInPresentAPI[i-n-1]`

**Comment by aggieNick02** [comment]
The 14-16 ms time in Q1 caught my eye - you might take a look at this, and make sure you are passing the process name in to presentmon: https://github.com/GameTechDev/PresentMon/issues/14

**Comment by DevTechProfile** [comment]
Here's a Doom Eternal capture file: https://www.dropbox.com/s/3zb779iz9p7mm5l/PresentMon-2021-07-05T223722.csv?dl=0

msUntilDisplayed is up to 14ns and higher.

**Comment by aggieNick02** [comment]
I'm not an expert at analyzing the files, but it definitely looks like there is a framerate limiter, whether internal or external, on the doom run. The between display change times are fairly regular at 8.3ms or 120Hz for the frames that are not dropped. 

Not sure about the msUntilDisplayed, agreed those are huge. MIght be worth figuring out the framerate limiting first though. If the framerate limiting is holding the earliest frame of a batch to display it, that would give a huge msUntilDisplayed, but the numbers in that CSV don't seem to even match that theory.

**Comment by DevTechProfile** [comment]
VSync was definitely off. I've measured Strange Brigade which implements both APIs, DX12 and Vulkan.

DX12, G-Sync off, VSync off, no frame limiter: https://www.dropbox.com/s/rv4504srgo4b5bw/DX12_GSync_off-2021-07-06T204830.csv?dl=0

msUntilDisplayed ~4-10ms

Vulkan, G-Sync off, VSync off, no frame limiter: https://www.dropbox.com/s/1uil4vm4sbp8pe7/Vulkan_GSync_off-2021-07-06T204444.csv?dl=0

msUntilDisplayed ~22-25ms

**Comment by aggieNick02** [comment]
V-Sync may be off in one place, but I'm pretty confident it is on somewhere, whether at the driver level or in the game itself. All of the frames in the Doom Vulkan and Strang Brigade Vulkan runs have "Allows Tearing" as 0, which means vsync is on. The fact that only one frame every 1/120th of a second is displayed confirms it as well.

**Comment by DevTechProfile** [comment]
Reset driver to defaults and then VSync forced off (in via driver).

Vulkan, G-Sync off, VSync forced off (driver), no frame limiter: https://www.dropbox.com/s/e1qrjaxo232etjf/VSync_forced_off-2021-07-06T212845.csv?dl=0

msUntilDisplayed ~22-25ms

**Comment by aggieNick02** [comment]
So vsync is not the right term for what you are seeing. Frames are being dropped and no tearing is occurring at all. But the engine is running as fast as it can, unlike proper vsync.

My best guess is that the games when running under Vulkan are not actually running fullscreen, but instead are running borderless windowed. I've got some runs of ashes of the singularity under Vulkan that don't have the issue you are seeing, and list the PresentMode as LegacyFlip. Under borderless windowed, some frames just aren't displayed, and there is a substantial delay for the frames that aren't, which explains the really high msUntilDisplayed as well as the fact that only a frame per refresh makes it through.

Some related articles:
https://www.reddit.com/r/Doom/comments/4siw8x/switching_to_vulkan_only_allows_for_borderless/
https://www.reddit.com/r/pcmasterrace/comments/7vpqni/input_lag_tests_tldr_dont_use_borderless_windowed/
https://steamcommunity.com/app/379720/discussions/0/358417461603764422/?l=german&ctp=2

**Comment by Taxxor90** [comment]
IIRC Windows 10 sould have some sort of sync mechanism that is used when playing in windowed mode. And since borderless fullscreen is essentially windowed, this might be the same case.

**Comment by DevTechProfile** [comment]
> My best guess is that the games when running under Vulkan are not actually running fullscreen, but instead are running borderless windowed.

Yes, that was the problem. Switching over to fullscreen mode leads to 3-4ms msUntilDisplayed numbers.

**Comment by JeffersonMontgomery-Intel** [comment]
Q1) upper bound looks good but strictly speaking it could be as low as `msUntilDisplayed[i]` if the application latches input late in its frame.  e.g., `msUntilDisplayed[i]` would correspond to input collected immediately before calling Present().

Q2) I believe the numbers, and your discussion here, is correct.  If there is still something suspicious I could help look closer if you can share an ETL.

Q3) Use the same equations as Q1 for frames that are displayed.  Don't compute a latency for dropped frames, and exclude them from any statistics you are computing, such as averages etc.

**Comment by Taxxor90** [comment]
> Don't compute a latency for dropped frames

Our current model had this equation be done for any frame that wasn't dropped but because we're going through the entire list of frametimes, for things like frameTimes[i - 1]  or  inPresentAPITimes[i - 2], these indexes could still be a frame that was dropped.
So the list of frametimes to do these equations with should better be an already filtered list without any dropped frames? 

The idea of @aggieNick02 that dropped frames add to the latency because they are adding additional msBetweenPresents times between two rendered frames sounded quite logical to me.

> and exclude them from any statistics you are computing, such as averages etc.

By statistics you mean input lag statistics or also performance statistics like average frametimes/FPS ?

**Comment by JeffersonMontgomery-Intel** [comment]
>  the equation is only done for frames that are displayed, but could include variables of dropped frames(in this case the inPresentAPITimes[i-1] and in case of the upper bound also the frametime[i-1]).  Is this what you mean, or do you mean the list of frametimes to do these equations with should better be an already filtered list without any dropped frames?

That code looks fine.  msBetweenPresents and msInPresentAPI are valid for dropped frames.

> appMissedCount is only there to make sure that the index of msInPresentAPI is always one that wasn't a dropped

Dropped frames still run the cpu/gpu work (including sampling input etc.) so I don't think this is correct: just use the same equation.

> By statistics you mean input lag statistics

Yes, that's what I was talking about and looks like you're already doing that.  But, for completeness, you also don't want to include dropped frames in anything display related such as an average of displayed duration (msBetweenDisplayChange).

**Comment by aggieNick02** [comment]
All of this got me a bit confused and I'm trying to wrap my head around it again, and found myself starting at the first latency approximation on the main page.
`LatencyMs =~ msBetweenPresents + msUntilDisplayed - previous(msInPresentAPI)`
That is an approximation for latency if the input is sampled right after present. It is the time between when input is sampled and when that sampled input, ignoring display hardware latency, is realized in the on-screen image.
But the latency between when input *occurs* and when it is realized on screen can be more, depending on when the input happens versus when it is sampled, correct? Let's call ps[k] the time when present k begins and pf[k] the time when present k ends. A button press captured immediately after pf[i-1] may have occurred anytime between pf[i-2] and pf[i-1]. So the latency for present i, which captures its input at time pf[i-1], can be as much as 
`ps[i]-pf[i-2] + msUntilDisplayed[i] =`
`ps[i]-ps[i-1]+ps[i-1]-ps[i-2]+ps[i-2]-pf[i-2]+msUntilDisplayed[i] =`
`msBetweenPresent[i]+msBetweenPresent[i-1] + msUntilDisplayed[i]-msInPresent[i-2]`

This is precisely the upper bound @DevTechProfile had in their initial post. And the lower bound is a good lower bound as it is precisely the button press to screen delay if the button press happens right when input is captured, for the scenario with input capture right after present.

It is true that if the application samples input right before present, then the button to screen delay could be even lower. But that is assuming a different input sampling strategy (right before present vs right after), so then the upper bound for that input sampling strategy needs to be reformulated as well. It would be
`msBetweenPresenst[i] + msUntilDisplayed[i]`

In summary, if sampling happens immediately post-present, we've got
Lower bound: `msBetweenPresents[i] + msUntilDisplayed[i] - msInPresent[i-1]`
Upper bound: `msBetweenPresents[i] + msBetweenPresent[i-1] + msUntilDisplayed[i] - msInPresent[i-2]`

And if sampling happens immediately pre-present, we've got
Lower bound: `msUntilDisplayed[i]`
Upper bound: `msBetweenPresents[i] + msUntilDisplayed[i]`

And this checks out too, because the difference between the upper bounds is msBetweenPresent[i-1]-msInPresent[i-2], and the difference between the lower bounds is msBetweenPresents[i] - msInPresent[i-1], or precisely the amount of time between when  a present finishes and the next present begins.

Does that make sense? Am I thinking about this right?

**Comment by aggieNick02** [comment]
> Q3) Use the same equations as Q1 for frames that are displayed. Don't compute a latency for dropped frames, and exclude them from any statistics you are computing, such as averages etc.

@JeffersonMontgomery-Intel 
I feel like trying to talk about latency approximations when we've got dropped frames is really ill-defined. When there are dropped frames, the button press to display latency goes up. For example, let's assume post-present sampling, and say frame 10 is displayed but 11 and 12 are dropped and 13 is displayed.

The input sampled right after frame 10's present doesn't get realized on screen until frame 13 is displayed. The engine may incorporate it in the rendered but dropped frames 11 and 12, but those never appear, and don't have an associated msUntilDisplayed either. Frames 11 and 12 could have been problematic and yielded huge latency had they been sent to the screen, and just because they are dropped doesn't mean they should get a pass and be excluded.

But at the same time it's hard to do anything meaningful with them - while the latency has increased, it is because the display hardware can't handle the generated framerate and the game is running in a mode that doesn't allow tearing. A different refresh rate monitor would change which frames are dropped and what kind of latency is observed in the real world.

So we've got these dropped frames that we can't use to talk about latency, but it is also not right to ignore them. My inclination is to say you can't really talk about display-independent latency approximations at all when you've got dropped frames.

**Comment by aggieNick02** [comment]
>Yes, that's what I was talking about and looks like you're already doing that. But, for completeness, you also don't want to include dropped frames in anything display related such as an average of displayed duration (msBetweenDisplayChange).

I kinda want to go a step further and say that if your run includes dropped frames, trying to talk about any display related statistics can be ill-defined or pointless. For example, if you exclude the dropped frames from your vulkan run, your average of displayed duration is 8.333ms . But that number's not really interesting at all. For *any* run where tearing is not allowed, your average displayed duration is going to be the one divided by the refresh rate of your monitor.

**Comment by Taxxor90** [comment]
> Frames 11 and 12 could have been problematic and yielded huge latency had they been sent to the screen, and just because they are dropped doesn't mean they should get a pass and be excluded.

I‘d say in case of the Vulkan example they were dropped because they were too fast to get displayed, so there is no latency they could’ve added because of they were slower they wouldn’t have been dropped

**Comment by aggieNick02** [comment]
> > Frames 11 and 12 could have been problematic and yielded huge latency had they been sent to the screen, and just because they are dropped doesn't mean they should get a pass and be excluded.
> 
> I‘d say in case of the Vulkan example they were dropped because they were too fast to get displayed, so there is no latency they could’ve added because if they were slower they wouldn’t have been dropped

In the Vulkan runs attached, I get what you're saying, but that doesn't mean you can just ignore the csv entries. If you do that you're ignoring portions of the input window that are not reflected until the non-dropped frame is displayed.

Also, you could have runs where a problematic frame takes a really long time (multiples of a screen refresh) but ends up getting dropped because after it completes another frame completes quickly and is displayed instead. You can't just ignore all the time eaten up by that problematic frame just because it was dropped.

**Comment by Taxxor90** [comment]
> You can't just ignore all the time eaten up by that problematic frame just because it was dropped.


But msBetweenPresent[i-1] in the upper bound equation would account for that dropped frame

**Comment by DevTechProfile** [comment]
@JeffersonMontgomery-Intel It would be nice if you could something about @aggieNick02's arguments. 

Someone from the DigitialFoundry team asked me yesterday: "How do you calculate it and can you be sure it's accurate?"

Is it somehow possible to quantify the error? What we have observed is that the difference between upper and lower bound can be about 50% (relative to lower bound).

**Comment by JeffersonMontgomery-Intel** [comment]
> But the latency between when input occurs and when it is realized on screen can be more, depending on when the input happens versus when it is sampled [...] Does that make sense? Am I thinking about this right?

Yes, good summary!  That is why CapFrameX introduced the upper/lower bounds on latency. 

> When there are dropped frames, the button press to display latency goes up. For example, let's assume post-present sampling, and say frame 10 is displayed but 11 and 12 are dropped and 13 is displayed.  The input sampled right after frame 10's present doesn't get realized on screen until frame 13 is displayed.

Dropped frames are still fully processed/generated by the cpu/gpu, they just don't get displayed.  So, input sampled after frame 9 is shown when frame 10 is displayed. Input sampled after frame 10 is used to generate frame 10, but ultimately never displayed.  It's the input sampled after frame 12 that is used to generate frame 13 and is shown when frame 13 is displayed.

> Frames 11 and 12 could have been problematic and yielded huge latency had they been sent to the screen, and just because they are dropped doesn't mean they should get a pass and be excluded.

The cost of dropped frames shows up as higher display duration for the previous frame, and that's an important metric for animation quality.  So, it's not totally excluded... it just doesn't make sense to try and compute input->display latency for them, since they're never displayed.

> if you exclude the dropped frames from your vulkan run, your average of displayed duration is 8.333ms . But that number's not really interesting at all. For any run where tearing is not allowed, your average displayed duration is going to be the one divided by the refresh rate of your monitor.

With vsync, any individual frame's display duration will be a multiple of the refresh rate.  The average will only be 8.3ms if every single frame in the range took < 8.3ms to get generated.  If some of those took longer, then the frames before them will be displayed for 16.6ms and the average would be something in between.

This is why I use counts of frames that occurred at each refresh multiple for display duration of vsync'ed workloads (instead of average). e.g., for this scenario I'd say something like 83% of frames @ 120fps and 17% @ 60.

> Is it somehow possible to quantify the error? What we have observed is that the difference between upper and lower bound can be about 50% (relative to lower bound).

Yeah, the range of possibilities is pretty large without more information.  I've experimented with some ideas to improve this, and continue to work on it, but there is no better solution right now unless you can build or get access to one of the HW measuring tools.

**Comment by Wat00** [comment]
CXWorld/CapFrameX#182
